pkg/lib/cockpit.js

```diff
@@ -3499,24 +3501,23 @@ function factory() {
                 if (watch_channel)
                     return;
 
-                const opts = {
-                    payload: "fswatch1",
-                    path,
-                    superuser: base_channel_options.superuser,
-                };
-                watch_channel = cockpit.channel(opts);
-                watch_channel.addEventListener("message", function (event, message_string) {
-                    let message;
-                    try {
-                        message = JSON.parse(message_string);
-                    } catch (e) {
-                        message = null;
-                    }
-                    if (message && message.path == path && message.tag && message.tag != watch_tag) {
-                        if (options && options.read !== undefined && !options.read)
-                            fire_watch_callbacks(null, message.tag);
-                        else
-                            read();
+                watch_channel = new FsInfoClient(path, ["tag"], { superuser: base_channel_options.superuser });
+                watch_channel.on('change', (state) => {
+                    if (state.error || state.info) {
+                        // Behave like fsread1, not-found is not a fatal error
+                        if (state.error && state.error?.problem !== "not-found") {
+                            const error = new BasicError(state.error.problem, state.error.message);
+                            fire_watch_callbacks(null, "-", error);
+                        } else {
+                            const tag = state?.info?.tag || "-";
```

Comment:
Too much `?` here.  `state` is definitely non-null.  `info` might indeed be null in the problem=="not-found" case.  A missing `tag` does not indicate that the file is missing, though — it means that we couldn't `stat()` it.

So:
  - null state: not possible
  - null `info` (ie: error is set, and we know it's "not-found" because of above): tag of '-' is correct
  - null `tag`: we don't know what's going on, so probably do nothing

So the logic is something like

```js
    if (state.error) {
        if (state.error.problem === "not-found") {
            // file not found is equivalent to tag of "-"
            if (watch_tag != "-")
                fire_watch_callbacks(null, "-", null);
        } else {
            // other errors are reported as errors
            const error = new BasicError(state.error.problem, state.error.message);
            fire_watch_callbacks(null, null, error);
        )
    } else if (state.info && state.info.tag) {
        // otherwise, the file is present with the given tag
        if (state.info.tag !== watch_tag) {
            if (options?.read)
                read();
            else
                fire_watch_callbacks(null, tag);
        }
    }
```